Utilities/Mergers/src/Merger.cxx

```diff
@@ -0,0 +1,284 @@
+    case MergingMode::Binwise:
+    case MergingMode::UseInterfaceFunction: {
+
+      size_t i = 0;
+      if (!mMergedObjects) {
+        for (; i < mCache.size(); i++) {
+          if (!mCache[i].deque.empty()) {
+            mMergedObjects.reset(mCache[i].deque[0].obj->Clone());
+            mCache.setMerged(i, 0);
+            break;
+          }
+        }
+      }
+
+      if (!mMergedObjects) {
+        LOG(INFO) << "mergeCache(): The cache is empty, nothing to merge.";
+        return;
+      }
+
+      auto unpackedMergedObjects = unpackObjects(mMergedObjects.get());
+      std::vector<TObjArray> unpackedCollectionsOfObjects(unpackedMergedObjects.size());
+      // todo: unpack straight to TCollection?
+      for (; i < mCache.size(); i++) {
+        for (const auto& entry : mCache[i].deque) {
+          if (!entry.is_merged) {
+            auto unpackedCachedObjects = unpackObjects(entry.obj.get());
+            assert(unpackedMergedObjects.size() == unpackedCachedObjects.size());
+
+            for (int j = 0; j < unpackedCachedObjects.size(); j++) {
+              unpackedCollectionsOfObjects[j].Add(unpackedCachedObjects[j]);
+            }
+          }
+        }
+      }
+
+      for (int k = 0; k < unpackedMergedObjects.size(); k++) {
+
+        TObject* mergedObject = unpackedMergedObjects[k];
+        const char* className = mergedObject->ClassName();
+        Long64_t errorCode = 0;
+
+        //todo: investigate -NOCHECK flag for histogram merging
+
+        if (mConfig.mergingMode.value == MergingMode::UseInterfaceFunction) {
+          errorCode = reinterpret_cast<MergeInterface*>(mergedObject)->merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strncmp(className, "TH1", 3) == 0) {
+          errorCode = reinterpret_cast<TH1*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strncmp(className, "TH2", 3) == 0) {
+          errorCode = reinterpret_cast<TH2*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strncmp(className, "TH3", 3) == 0) {
+          errorCode = reinterpret_cast<TH3*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strncmp(className, "THn", 3) == 0) {
+          errorCode = reinterpret_cast<THn*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strncmp(className, "THnSparse", 8) == 0) {
+          errorCode = reinterpret_cast<THnSparse*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else if (strcmp(className, "TTree") == 0) {
+          errorCode = reinterpret_cast<TTree*>(mergedObject)->Merge(&unpackedCollectionsOfObjects[k]);
+        } else {
+          //          LOG(ERROR) << "Object with type " << className << " is not one of mergeable type.";
+          throw std::runtime_error("Object with type '" + std::string(className) + "' is not one of mergeable type.");
+          // todo: maybe it is fine to just overwrite?
+        }
```

Comment:
based on run2 experience I would suggest to introduce an interface class for mergeable objects in o2 - it would simplify maintenance of merging code (would become generic) and it would keep the control over custom merging with user code - a bit dangerous to let people execute their code, maybe, but with testing could be no problem.

Reply:
I'd second that. Both on general grounds and also because in Run2 MCH was known to have a mergeable custom object as part of its QC ;-) (and that may not drastically change for Run3)

Reply:
Isn't this exactly what you mean?
https://github.com/AliceO2Group/AliceO2/blob/f91c6d84b29795a29ff0be55330c547c9bde6697/Utilities/Mergers/include/Mergers/MergeInterface.h

Reply:
possibly, but then could you please make the class abstract?

Reply:
Yes, I should have done that.